internal/pkg/cosign/payload/payload.go

```diff
@@ -32,46 +35,135 @@ type payloadSigner struct {
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	pkOpts := []signature.PublicKeyOption{signatureoptions.WithContext(ctx)}
 	pkOpts = append(pkOpts, ps.publicKeyProviderOpts...)
-	pk, err := ps.payloadSigner.PublicKey(pkOpts...)
+	pk, err = ps.payloadSigner.PublicKey(pkOpts...)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return payloadBytes, sig, pk, nil
+}
+
+// Sign implements `Signer`
+func (ps *payloadSigner) Sign(ctx context.Context, payload io.Reader) (oci.Signature, crypto.PublicKey, error) {
+	payloadBytes, sig, pk, err := ps.signPayload(ctx, payload)
 	if err != nil {
 		return nil, nil, err
 	}
 
 	b64sig := base64.StdEncoding.EncodeToString(sig)
-	ociSig, err := static.NewSignature(payloadBytes, b64sig)
+
+	var opts []static.Option
+	if len(ps.certPEM) > 0 {
+		opts = []static.Option{static.WithCertChain(ps.certPEM, ps.chainPEM)}
+	}
+	ociSig, err := static.NewSignature(payloadBytes, b64sig, opts...)
 	if err != nil {
 		return nil, nil, err
 	}
 
 	return ociSig, pk, nil
 }
 
-// NewSigner returns a `cosign.Signer` uses the given `signature.Signer` to sign the requested payload, then returns the signature, the public key associated with it, the signed payload
-func NewSigner(s signature.Signer,
+// Attest implements `cosign.Attestor`
+func (pa *payloadAttestor) Attest(ctx context.Context, payload io.Reader) (oci.Signature, crypto.PublicKey, error) {
+	p, err := io.ReadAll(payload)
+	if err != nil {
+		return nil, nil, err
+	}
+	pae := dsse.PAE(pa.payloadType, string(p))
+
+	_, sig, pk, err := pa.signPayload(ctx, bytes.NewReader(pae))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	envelope := dsse.Envelope{
+		PayloadType: pa.payloadType,
+		Payload:     base64.StdEncoding.EncodeToString(p),
+		Signatures: []dsse.Signature{
+			{
+				Sig: base64.StdEncoding.EncodeToString(sig),
+			},
+		},
+	}
+
+	envelopeJSON, err := json.Marshal(envelope)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	opts := []static.Option{static.WithLayerMediaType(types.DssePayloadType)}
+
+	att, err := static.NewAttestation(envelopeJSON, opts...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return att, pk, nil
+}
+
+func newSigner(s signature.Signer,
 	sOpts []signature.SignOption,
-	pkOpts []signature.PublicKeyOption) cosign.Signer {
-	return &payloadSigner{
+	pkOpts []signature.PublicKeyOption,
+	certPEM, chainPEM []byte) payloadSigner {
+	return payloadSigner{
 		payloadSigner:         s,
 		payloadSignerOpts:     sOpts,
 		publicKeyProviderOpts: pkOpts,
+
+		certPEM:  certPEM,
+		chainPEM: chainPEM,
+	}
+}
+
+// NewSigner returns a `cosign.Signer` which uses the given `signature.Signer` to sign requested payloads.
+// The cert and chain, if provided, will be included in returned `oci.Signature`s.
+func NewSigner(s signature.Signer,
+	sOpts []signature.SignOption,
+	pkOpts []signature.PublicKeyOption,
+	certPEM, chainPEM []byte) cosign.Signer {
```

Comment:
This is an enormous signature.  We are growing `payload.NewSigner` into the monstrosity that was previously in `cmd`.  The whole point of the `fulcio` package is to have that encapsulate the variants that deal with chain-based signing, so I think it's time for `fulcio.NewSigner` to stop wrapping `payload.NewSigner` and take on some of this logic itself.

Reply:
I can split the `cert` and `chain` into their own layer. Alternatively, we could expose the struct directly.

>  I think it's time for fulcio.NewSigner to stop wrapping payload.NewSigner and take on some of this logic itself

I don't agree, since I don't think we want to limit the Fulcio path to keyless